semantic-convention-utils/src/main/java/org/hypertrace/semantic/convention/utils/http/HttpSemanticConventionUtils.java

```diff
@@ -196,4 +257,257 @@ public static boolean isValidUrl(String url) {
+        if (StringUtils.isNotBlank(s) && s.startsWith(SLASH)) {
+          return getPathFromUrlObject(s).map(str -> removeTrailingSlash(str));
+        }
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpMethod(Event event) {
+
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String method : METHOD_ATTRIBUTES) {
+      if (attributeValueMap.get(method) != null
+          && !StringUtils.isBlank(attributeValueMap.get(method).getValue())) {
+        return Optional.of(attributeValueMap.get(method).getValue());
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpScheme(Event event) {
+
+    Optional<String> url = getHttpUrlUtil(event);
+    if (url.isPresent() && isAbsoluteUrl(url.get())) {
+      try {
+        return Optional.of(getNormalizedUrl(url.get()).getProtocol());
+      } catch (MalformedURLException e) {
+        e.printStackTrace();
+      }
+    }
+    return getHttpSchemeUtil(event);
+  }
+
+  private static Optional<String> getHttpSchemeUtil(Event event) {
+
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String scheme : SCHEME_ATTRIBUTES) {
+      if (attributeValueMap.get(scheme) != null
+          && !StringUtils.isEmpty(attributeValueMap.get(scheme).getValue())) {
+        return Optional.of(attributeValueMap.get(scheme).getValue());
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getHttpUrl(Event event) {
+    Optional<String> url = getHttpUrlUtil(event);
+    if (url.isPresent() && !isAbsoluteUrl(url.get())) {
+      return Optional.empty();
+    }
+    return getHttpUrlUtil(event);
+  }
+
+  //  input url to populateurlparts
+  private static Optional<String> getHttpUrlUtil(Event event) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String url : FULL_URL_ATTRIBUTES) {
+      if (attributeValueMap.get(url) != null) {
+        String urlVal = attributeValueMap.get(url).getValue();
+        if (!StringUtils.isBlank(urlVal) && isValidUrl(urlVal)) {
+          if (isAbsoluteUrl(urlVal)) {
+            return Optional.of(urlVal);
+          } else {
+            Optional<String> url2 = getHttpUrlForOTelFormat(attributeValueMap);
+            return url2.isPresent() ? url2 : Optional.of(urlVal);
+          }
+        }
+      }
+    }
+
+    return getHttpUrlForOTelFormat(attributeValueMap);
```

Comment:
this return is same as the else block, why not club the conditions together?

Reply:
its outside the for loop

Reply:
Ah right, why is the call `getHttpUrlForOTelFormat` inside the loop, it isn't dependant on the loop var

Reply:
This is in reference to this function:
https://github.com/hypertrace/hypertrace-ingester/blob/0261e7eef2641b0576c6cba05a215e976b6676f5/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L567

Outside for loop is if _requestBuilder.hasUrl()_ return false
The other one is for when _isAbsoluteUrl(requestBuilder.getUrl())_ return false

Reply:
This doesn't seem correct. 
Lets say for the first item in the list of attributes it is a valid url but not an absolute url, you would return otel format url or the original string itself without looking at rest of the attributes in the list

Reply:
Isn't this happening in the original code as well ? Once the url is set, other keys are not looked at

https://github.com/hypertrace/hypertrace-ingester/blob/0261e7eef2641b0576c6cba05a215e976b6676f5/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/HttpFieldsGenerator.java#L424

Reply:
> Isn't this happening in the original code as well ? Once the url is set, other keys are not looked at

In the existing code, url is set using one of the attributes in the `FULL_URL_ATTRIBUTES` list, if this set url is not an absolute url then it is conditionally overridden in this method `maybeSetHttpUrlForOtelFormat`